### PR TITLE
Added BUILD rules for C++ wasm compilation with Emscripten

### PR DESCRIPTION
**Pull request by csilvestrini**
Not actually used anywhere yet.

This commit is an example of how it is used: https://github.com/PolymerLabs/arcs/pull/3525/commits/62dfc71af4d4457554c9d5dc3aeda22fd52f1501

### CODE REVIEW

File: build_defs/emscripten/build_defs.bzl [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        transitive = [dep[WasmLibInfo].hdrs for dep in deps],
+    )
+    return src_depset, hdr_depset
+
+def _cc_wasm_binary(ctx):
+    args = ctx.actions.args()
+    args.add_all(_emscripten_args)
+
+    # For workspace-relative #includes.
+    args.add("-I", ".")
```
Comment (mykmartin) [inline build_defs/emscripten/build_defs.bzl]: Could this just be in the default list?

**Comment by cromwellian** [comment]
Was this forked from somewhere, or did you write it from scratch?

**Comment by cromwellian** [comment]
Just for reference, the bazel folks seem to like using toolchains to do this, see

https://github.com/ribrdb/rules_emscripten
https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html

Which includes WASM example.

The reason I mention this is that the internal version of this rule using the Bazel 'transition' feature to swap toolchains to support wasm (https://docs.bazel.build/versions/master/skylark/lib/transition.html)

Not that I object to this implementation, but eventually CC and Kotlin will settle on the transition technique I think.

File: build_defs/emscripten/build_defs.bzl (hunk 1 above)
Comment (csilvestrini) [reply build_defs/emscripten/build_defs.bzl]: Yep, done!

**Comment by csilvestrini** [comment]
> Was this forked from somewhere, or did you write it from scratch?

It was inspired by the emsdk-npm module, which does basically the same thing but for npm.

> Just for reference, the bazel folks seem to like using toolchains to do this, see
> https://github.com/ribrdb/rules_emscripten
> https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html
> Which includes WASM example.

I saw the toolchain tutorial already, but it was for asmjs, not wasm. I had quite a bit of trouble trying to update it to work with wasm (the tutorial just seems out of date in general). The rules_emscripten repo looks promising though.

> The reason I mention this is that the internal version of this rule using the Bazel 'transition' feature to swap toolchains to support wasm (https://docs.bazel.build/versions/master/skylark/lib/transition.html)
> Not that I object to this implementation, but eventually CC and Kotlin will settle on the transition technique I think.

Replacing this with transitions SGTM. I just wanted to get something working, and this was the only way I could see to move forward. Another part of the problem is that they seem to change the wasm-js API bridge in pretty much every emscripten release. The generated symbols were changing in every which way as I tried out different versions. I ended up having to use the same emsdk version which Myk had pegged to in emsdk-npm to get the same symbol names. Even using what was the emscripten release tarballs corresponding to allegedly the same version number was giving different output. It seems incredibly fragile to me.

In summary, using transitions is probably the way to go, I'd be very happy for someone to swap out the implementation, provided that all our existing C++ wasm code keeps working.
